x-pack/elastic-agent/pkg/agent/cmd/container.go

```diff
@@ -112,15 +117,97 @@ all the above actions will be skipped, because the Elastic Agent has already bee
 occurs on every start of the container set FLEET_FORCE to 1.
 `,
 		Run: func(c *cobra.Command, args []string) {
-			if err := containerCmd(streams, c, flags, args); err != nil {
-				fmt.Fprintf(streams.Err, "Error: %v\n", err)
+			var err error
+			if _, cloud := os.LookupEnv("ELASTIC_AGENT_CLOUD"); cloud {
+				err = containerCloudCmd(streams, c, flags, args)
+			} else {
+				err = containerCmd(streams, c, flags, defaultAccessConfig())
+			}
+			if err != nil {
+				logError(streams, err)
 				os.Exit(1)
 			}
 		},
 	}
+	return &cmd
+}
+
+func logError(streams *cli.IOStreams, err error) {
+	fmt.Fprintf(streams.Err, "Error: %v\n", err)
 }
 
-func containerCmd(streams *cli.IOStreams, cmd *cobra.Command, flags *globalFlags, args []string) error {
+func logInfo(streams *cli.IOStreams, msg string) {
+	fmt.Fprintln(streams.Out, msg)
+}
+
+func containerCloudCmd(streams *cli.IOStreams, cmd *cobra.Command, flags *globalFlags, args []string) error {
+	logInfo(streams, "Elastic Agent container in cloud mode")
+	// sync main process and apm-server legacy process
+	var wg sync.WaitGroup
+	wg.Add(1) // main process (always running)
+	mainProc, err := os.FindProcess(os.Getpid())
+	if err != nil {
+		return errors.New(err, "finding current process")
+	}
+	var apmProc *process.Info
+	// run legacy APM Server as a daemon; send termination signal
+	// to the main process if the daemon is stopped
+	apmPath := os.Getenv("APM_SERVER_PATH")
+	if apmPath != "" {
+		apmProc, err = runLegacyAPMServer(streams, apmPath, args)
+		if err != nil {
+			return errors.New(err, "starting legacy apm-server")
+		}
+		logInfo(streams, "Legacy apm-server daemon started.")
+		wg.Add(1) // apm-server legacy process
+		go func() {
+			if err := func() error {
+				apmProcState, err := apmProc.Process.Wait()
+				if err != nil {
+					return err
```

Comment:
When I read this code first, I was thinking wg.Done() will never be called but realised it is inside a function. To make it more clear wg.Done() is always called, maybe we could use a defer on line 165? Same on line 186

Reply:
`wg.Done()` needs to be called inside the go routine, as it is used for syncing when the go routine finishes. I specifically put it before sending the termination signal for the agent. If the agent container already finished, the `wg.Wait` will trigger before a signal to terminate the main process is sent. It is just a minor detail, without big consequences, but it makes more sense to call `wg.Done` where it is called right now.